### PR TITLE
FortiManager integration

### PR DESCRIPTION
**Pull request by roysagi**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/16318

## Description
new FortiManager integration

TODO:
 - [x] Check all the methods. - done for most - missing some for technical reasons.
    - [x] Policy related commands do not work - might be a wrong API call or wrong use, asked a question on Forti forum.
    - [x] Dynamic mapping - @Arsenikr is not sure if needed on second glance, skipping for now.
    - [x] Installation process related commands - @Arsenikr - should get back to me on this.
 - [x] Add description.
 - [x] Add unit tests and test playbook. - can be improved once the open checks are done
 - [x] Generate docs. - To be done after open checks mission

### CODE REVIEW

**Review by ronykoz — COMMENTED** [review]
Please update the logo of the integration

**Comment by ronykoz** [comment]
@teizenman let's start going over the code

File: Packs/FortiManager/Integrations/FortiManager/FortiManager_description.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## Hello World
+- This section explains how to configure the instance of HelloWorld in Cortex XSOAR.
+- You can use the following API Key: `43ea9b2d-4998-43a6-ae91-aba62a26868c`
```
Comment (ronykoz) [inline Packs/FortiManager/Integrations/FortiManager/FortiManager_description.md:3]: let's update this

File: Packs/FortiManager/Integrations/FortiManager/FortiManager.py [hunk 2]
```diff
@@ -0,0 +1,728 @@
...
+            url_suffix='jsonrpc',
+            json_data=body
+        )
+        return response
+
+    def fortimanager_api_call(self, method, url, data=None, range_info=None, other_params=None):
+        response = self.fortimanager_http_request(method, url, data)
+
+        # catch session token expiration - fetch new token and retry
+        if response.get('result')[0].get('status').get('code') != -11:
```
Comment (ronykoz) [inline Packs/FortiManager/Integrations/FortiManager/FortiManager.py:76]:
```suggestion
        if response.get('result')[0].get('status', {}).get('code') != -11:
```

File: Packs/FortiManager/Integrations/FortiManager/FortiManager.py [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+
+        if response.get('result')[0].get('status').get('code') != 0:
+            raise DemistoException(response.get('result')[0].get('status').get('message'))
+
+        return response.get('result')[0].get('data')
+
+
+def get_global_or_adom(client, args):
+    adom = args.get('adom') if args.get('adom') else client.adom
+    if adom == GLOBAL_VAR:
```
Comment (ronykoz) [inline Packs/FortiManager/Integrations/FortiManager/FortiManager.py:89]: add a comment here

File: Packs/FortiManager/Integrations/FortiManager/FortiManager.py [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+    list_range = []
+
+    if first_index:
+        list_range.append(first_index)
+
+    if last_index:
+        list_range.append(list_range)
+
+    if list_range:
+        return [list_range]
```
Comment (ronykoz) [inline Packs/FortiManager/Integrations/FortiManager/FortiManager.py:118]: list of list?

**Review by ronykoz — COMMENTED** [review]
Let's update the image 
There are many commands that don't return outputs, is that intentional?

File: Packs/FortiManager/Integrations/FortiManager/FortiManager_description.md (hunk 1 above)
Comment (roysagi) [reply Packs/FortiManager/Integrations/FortiManager/FortiManager_description.md]: done

File: Packs/FortiManager/Integrations/FortiManager/FortiManager.py (hunk 3 above)
Comment (roysagi) [reply Packs/FortiManager/Integrations/FortiManager/FortiManager.py:101]: done

File: Packs/FortiManager/TestPlaybooks/Test-Fortimanager.yml [hunk 5]
```diff
@@ -0,0 +1,1995 @@
+id: 3da36d51-3cdf-4120-882a-cee03b038b89
+version: 7
```
Comment (ronykoz) [inline Packs/FortiManager/TestPlaybooks/Test-Fortimanager.yml:2]: update

File: Packs/FortiManager/Integrations/FortiManager/FortiManager_description.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
+## FortiManager
+- FortiManager is a single console central management system that manages Fortinet Devices.
+- You can use FortiManger integration to:
+  - View, create, update and delete Addresses, Address-Groups, Services, Service-Groups, Policies and Policy-Packages found in FortiManger.
```
Comment (ronykoz) [inline Packs/FortiManager/Integrations/FortiManager/FortiManager_description.md:4]: maybe add an explanation for getting the creds

**Review by ronykoz — APPROVED** [review]
I would suggest going over the outputs prior to assigning Alex

File: Packs/FortiManager/Integrations/FortiManager/FortiManager_description.md (hunk 6 above)
Comment (roysagi) [reply Packs/FortiManager/Integrations/FortiManager/FortiManager_description.md:4]: Not sure it is necessary, it's just the username and password for FortiManager.

**Comment by roysagi** [comment]
> Let's update the image
> There are many commands that don't return outputs, is that intentional?

Yes, A lot of the commands do not have any useful outputs in their response.

**Comment by roysagi** [comment]
@kirbles19 - if you could please check the following: 
1) [the yml file](https://github.com/demisto/content/blob/fortimanager_integration/Packs/FortiManager/Integrations/FortiManager/FortiManager.yml)
2) [the description file](https://github.com/demisto/content/blob/fortimanager_integration/Packs/FortiManager/Integrations/FortiManager/FortiManager_description.md)
3) [the readme file](https://github.com/demisto/content/blob/fortimanager_integration/Packs/FortiManager/Integrations/FortiManager/README.md)

**Comment by ShirleyDenkberg** [comment]
@ronykoz As agreed, only a very basic doc review was performed.  This integration was not tested.  The README should be regenerated.

**Comment by roysagi** [comment]
@ronykoz @Arsenikr - Waiting for some feedback from the customer about the integration and then we merge this.

**Comment by roysagi** [comment]
@kirbles19 @ShirleyDenkberg - If you can please go through the newly added `Required Permissions` section in the readme. Thank you!

**Comment by ShirleyDenkberg** [comment]
@ronykoz @teizenman Doc review done.

**Pull request merged**